dynamic_programming/subset_generation.py

```diff
@@ -1,44 +1,52 @@
-    arr[]  ---> Input Array
-    data[] ---> Temporary array to store current combination
-    start & end ---> Staring and Ending indexes in arr[]
-    index  ---> Current index in data[]
-    r ---> Size of a combination to be printed
+    Compute n-element combinations from a given list using dynamic programming.
+    Args:
+        elements (list): The list of elements from which combinations will be generated.
+        n (int): The number of elements in each combination.
+    Returns:
+        list: A list of tuples, each representing a combination of n elements.
+        >>> subset_combinations_dp(elements=[10, 20, 30, 40], n=2)
+        [(10, 20), (10, 30), (20, 30), (10, 40), (20, 40), (30, 40)]
+        >>> subset_combinations_dp(elements=[1, 2, 3], n=1)
+        [(1,), (2,), (3,)]
+        >>> subset_combinations_dp(elements=[1, 2, 3], n=3)
+        [(1, 2, 3)]
+        >>> subset_combinations_dp(elements=[42], n=1)
+        [(42,)]
+        >>> subset_combinations_dp(elements=[6, 7, 8, 9], n=4)
+        [(6, 7, 8, 9)]
+        >>> subset_combinations_dp(elements=[10, 20, 30, 40, 50], n=0)
+        [()]
+        >>> subset_combinations_dp(elements=[1, 2, 3, 4], n=2)
+        [(1, 2), (1, 3), (2, 3), (1, 4), (2, 4), (3, 4)]
+        >>> subset_combinations_dp(elements=[1, 'apple', 3.14], n=2)
+        [(1, 'apple'), (1, 3.14), ('apple', 3.14)]
+        >>> subset_combinations_dp(elements=['single'], n=0)
+        [()]
     """
-    if index == r:
-        for j in range(r):
-            print(data[j], end=" ")
-        print(" ")
-        return
-    #  When no more elements are there to put in data[]
-    if i >= n:
-        return
-    # current is included, put next at next location
-    data[index] = arr[i]
-    combination_util(arr, n, r, index + 1, data, i + 1)
-    # current is excluded, replace it with
-    # next (Note that i+1 is passed, but
-    # index is not changed)
-    combination_util(arr, n, r, index, data, i + 1)
-    # The main function that prints all combinations
-    # of size r in arr[] of size n. This function
-    # mainly uses combinationUtil()
-
-
-def print_combination(arr, n, r):
-    # A temporary array to store all combination one by one
-    data = [0] * r
-    # Print all combination using temporary array 'data[]'
-    combination_util(arr, n, r, 0, data, 0)
+    r = len(elements)
+
+    dp: list[list[tuple]] = [[] for _ in range(r + 1)]
+
+    dp[0].append(())
+
+    for i in range(1, r + 1):
+        for j in range(i, 0, -1):
+            for prev_combination in dp[j - 1]:
+                dp[j].append(tuple(prev_combination) + (elements[i - 1],))
+
+    combinations = dp[n]
+
+    return combinations
 
 
 if __name__ == "__main__":
-    # Driver code to check the function above
-    arr = [10, 20, 30, 40, 50]
-    print_combination(arr, len(arr), 3)
-    # This code is contributed by Ambuj sahu
+    print(f"{subset_combinations_dp(elements=[10, 20, 30, 40], n=2) = }")
```

Comment:
`subset_combinations_dp()` should work the same way as `itertools.combinations()`.  It should present the same sort order and raise the same exceptions.
```suggestion
    from itertools import combinations
    for items, n in (
        ([10, 20, 30, 40], 2), ([1, 2, 3], 1), ([1, 2, 3], 3), ([42], 1),
        ([6, 7, 8, 9], 4), ([10, 20, 30, 40, 50], 1), ([1, 2, 3, 4], 2),
        ([1, 'apple', 3.14], 2), (['single'], 0), ([], 9)
    ):
        actual = subset_combinations_dp(items, n)
        expected = set(combinations(items, n))
        assert actual == expected, f"items, n: {actual} != {expected}"


    print(f"{subset_combinations_dp(elements=[10, 20, 30, 40], n=2) = }")
```

Reply:
Now code giving the list in sort order

Reply:
`itertools.combinations([1, 'apple', 3.14], 2)`